File: app/src/main/java/com/giosis/util/qdrive/list/delivery/DeliveryFailedActivity.java
package com.giosis.util.qdrive.list.delivery;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore.Images;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.util.Size;
import android.view.KeyEvent;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.giosis.library.server.data.FailedCodeResult;
import com.giosis.util.qdrive.gps.GPSTrackerManager;
import com.giosis.util.qdrive.singapore.MyApplication;
import com.giosis.util.qdrive.singapore.OnServerEventListener;
import com.giosis.util.qdrive.singapore.R;
import com.giosis.util.qdrive.util.Camera2APIs;
import com.giosis.util.qdrive.util.DataUtil;
import com.giosis.util.qdrive.util.DisplayUtil;
import com.giosis.util.qdrive.util.MemoryStatus;
import com.giosis.util.qdrive.util.NetworkUtil;
import com.giosis.util.qdrive.util.PermissionActivity;
import com.giosis.util.qdrive.util.PermissionChecker;

import java.util.ArrayList;
import java.util.HashMap;

/***
 * @author eylee
 * @date 2017-06-23
 * Delivery failed -> Visit log 로 기능 개선
 *
 * @editor krm0219
 * 2020.12  Failed Reason 추가
 */
public class DeliveryFailedActivity extends AppCompatActivity implements Camera2APIs.Camera2Interface, TextureView.SurfaceTextureListener {
    private static final String TAG = "DeliveryFailedActivity";


    FrameLayout layout_top_back;
    TextView text_top_title;

    TextView text_sign_d_f_tracking_no;
    TextView text_sign_d_f_receiver;
    TextView text_sign_d_f_sender;

    // 2020.12
    RelativeLayout layout_sign_d_f_failed_reason;
    TextView text_sign_d_f_failed_reason;
    Spinner spinner_d_f_failed_reason;
    EditText edit_sign_d_f_memo;

    LinearLayout layout_sign_d_f_take_photo;
    LinearLayout layout_sign_d_f_gallery;
    TextureView texture_sign_d_f_preview;
    ImageView img_sign_d_f_preview_bg;
    ImageView img_sign_d_f_visit_log;
    Button btn_sign_d_f_save;


    //
    Context context;
    String opID = "";
    String officeCode = "";
    String deviceID = "";
    String mStrWaybillNo;

    // Camera & Gallery
    Camera2APIs camera2;
    String cameraId;
    private static final int RESULT_LOAD_IMAGE = 3;
    boolean isGalleryActivate = false;

    GPSTrackerManager gpsTrackerManager;
    boolean gpsEnable = false;
    double latitude = 0;
    double longitude = 0;

    boolean isPermissionTrue = false;
    private static final int PERMISSION_REQUEST_CODE = 1000;
    private static final String[] PERMISSIONS = new String[]{PermissionChecker.ACCESS_FINE_LOCATION, PermissionChecker.ACCESS_COARSE_LOCATION,
            PermissionChecker.READ_EXTERNAL_STORAGE, PermissionChecker.WRITE_EXTERNAL_STORAGE, PermissionChecker.CAMERA};

    ArrayList<FailedCodeResult.FailedCode> arrayList;
    ArrayList<String> failedCodeArrayList;
    ArrayAdapter<String> failedCodeArrayAdapter;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
                | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
        setContentView(R.layout.activity_delivery_visit_log);


        layout_top_back = findViewById(R.id.layout_top_back);
        text_top_title = findViewById(R.id.text_top_title);

        text_sign_d_f_tracking_no = findViewById(R.id.text_sign_d_f_tracking_no);
        text_sign_d_f_receiver = findViewById(R.id.text_sign_d_f_receiver);
        text_sign_d_f_sender = findViewById(R.id.text_sign_d_f_sender);

        layout_sign_d_f_failed_reason = findViewById(R.id.layout_sign_d_f_failed_reason);
        text_sign_d_f_failed_reason = findViewById(R.id.text_sign_d_f_failed_reason);
        spinner_d_f_failed_reason = findViewById(R.id.spinner_d_f_failed_reason);
        edit_sign_d_f_memo = findViewById(R.id.edit_sign_d_f_memo);

        layout_sign_d_f_take_photo = findViewById(R.id.layout_sign_d_f_take_photo);
        layout_sign_d_f_gallery = findViewById(R.id.layout_sign_d_f_gallery);
        texture_sign_d_f_preview = findViewById(R.id.texture_sign_d_f_preview);
        img_sign_d_f_preview_bg = findViewById(R.id.img_sign_d_f_preview_bg);
        img_sign_d_f_visit_log = findViewById(R.id.img_sign_d_f_visit_log);
        btn_sign_d_f_save = findViewById(R.id.btn_sign_d_f_save);


        //
        context = getApplicationContext();
        camera2 = new Camera2APIs(this);
        opID = MyApplication.preferences.getUserId();
        officeCode = MyApplication.preferences.getOfficeCode();
        deviceID = MyApplication.preferences.getDeviceUUID();

        String strTitle = getIntent().getStringExtra("title");
        mStrWaybillNo = getIntent().getStringExtra("waybillNo");
        String strReceiverName = getIntent().getStringExtra("receiverName");
        String strSenderName = getIntent().getStringExtra("senderName");


        text_top_title.setText(strTitle);
        text_sign_d_f_tracking_no.setText(mStrWaybillNo);
        text_sign_d_f_receiver.setText(strReceiverName);
        text_sign_d_f_sender.setText(strSenderName);
        DisplayUtil.setPreviewCamera(img_sign_d_f_preview_bg);


        // 202012.  Delivery Failed Reason 추가  (for Lazada)
        setFailedCode();

        layout_sign_d_f_failed_reason.setOnClickListener(v -> {

            spinner_d_f_failed_reason.setVisibility(View.INVISIBLE);
            spinner_d_f_failed_reason.performClick();
        });


        spinner_d_f_failed_reason.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> parentView, View arg1, int position, long arg3) {

                String selected_text = parentView.getItemAtPosition(position).toString();
                text_sign_d_f_failed_reason.setText(selected_text);
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {
            }
        });


        layout_top_back.setOnClickListener(view -> cancelSigning());


        layout_sign_d_f_take_photo.setOnClickListener(view -> {

            if (cameraId != null) {

                camera2.takePhoto(texture_sign_d_f_preview, img_sign_d_f_visit_log);
            } else {

                Toast.makeText(DeliveryFailedActivity.this, context.getResources().getString(R.string.msg_back_camera_required), Toast.LENGTH_SHORT).show();
            }
        });

        layout_sign_d_f_gallery.setOnClickListener(view -> getImageFromAlbum());


        btn_sign_d_f_save.setOnClickListener(v -> saveServerUploadSign());


        // Memo 입력제한
        edit_sign_d_f_memo.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                if (99 <= edit_sign_d_f_memo.length()) {
                    Toast.makeText(context, context.getResources().getText(R.string.msg_memo_too_long), Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });


        //
        PermissionChecker checker = new PermissionChecker(this);
        // 권한 여부 체크 (없으면 true, 있으면 false)
        if (checker.lacksPermissions(PERMISSIONS)) {

            isPermissionTrue = false;
            PermissionActivity.startActivityForResult(this, PERMISSION_REQUEST_CODE, PERMISSIONS);
            overridePendingTransition(0, 0);
        } else {

            isPermissionTrue = true;
        }
    }

    private void setFailedCode() {

        arrayList = DataUtil.getFailCode("D");

        if (arrayList == null) {

            AlertShow(context.getResources().getString(R.string.msg_failed_code_error));
        } else {

            failedCodeArrayList = new ArrayList<>();

            for (int i = 0; i < arrayList.size(); i++) {

                FailedCodeResult.FailedCode failedCode = arrayList.get(i);
                failedCodeArrayList.add(failedCode.getFailedString());
            }

            spinner_d_f_failed_reason.setPrompt(context.getResources().getString(R.string.text_failed_reason));
            failedCodeArrayAdapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_item, failedCodeArrayList);
            failedCodeArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spinner_d_f_failed_reason.setAdapter(failedCodeArrayAdapter);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (isPermissionTrue) {

            // Camera
            camera2 = new Camera2APIs(this);

            if (texture_sign_d_f_preview.isAvailable()) {

                openCamera();
            } else {

                texture_sign_d_f_preview.setSurfaceTextureListener(this);
            }

            // Location
            gpsTrackerManager = new GPSTrackerManager(context);
            gpsEnable = gpsTrackerManager.enableGPSSetting();

            if (gpsEnable && gpsTrackerManager != null) {

                gpsTrackerManager.GPSTrackerStart();
                latitude = gpsTrackerManager.getLatitude();
                longitude = gpsTrackerManager.getLongitude();

                Log.e("Location", TAG + " GPSTrackerManager onResume : " + latitude + "  " + longitude + "  ");
            } else {

                DataUtil.enableLocationSettings(DeliveryFailedActivity.this, context);
            }
        }
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        isGalleryActivate = false;

        if (requestCode == RESULT_LOAD_IMAGE && resultCode == RESULT_OK && null != data) {
            try {

                Uri selectedImageUri = data.getData();

                Bitmap selectedImage = Images.Media.getBitmap(getContentResolver(), selectedImageUri);
                Bitmap resizeBitmap = camera2.getResizeBitmap(selectedImage);
                img_sign_d_f_visit_log.setImageBitmap(resizeBitmap);
                img_sign_d_f_visit_log.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
                onResume();
            } catch (Exception e) {
                Log.e("eylee", e.toString());
                e.printStackTrace();
            }
        } else if (requestCode == PERMISSION_REQUEST_CODE) {   // permission
            if (resultCode == PermissionActivity.PERMISSIONS_GRANTED) {
                Log.e("Permission", TAG + "   onActivityResult  PERMISSIONS_GRANTED");

                isPermissionTrue = true;
            }
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        closeCamera();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        DataUtil.stopGPSManager(gpsTrackerManager);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        return super.onKeyDown(keyCode, event);
    }


    @Override
    public void onBackPressed() {
        cancelSigning();
    }

    public void cancelSigning() {

        new AlertDialog.Builder(this)
                .setMessage(R.string.msg_delivered_sign_cancel)
                .setPositiveButton(R.string.button_ok, (dialog, which) -> finish())
                .setNegativeButton(R.string.button_cancel, (dialog, which) -> dialog.dismiss()).show();
    }


    /*
     * 실시간 Upload 처리
     * 2016-10-04 eylee added
     * visit log 페이지에서 저장 버튼 눌렀을 때
     */
    public void saveServerUploadSign() {

        try {

            if (!NetworkUtil.isNetworkAvailable(this)) {

                AlertShow(context.getResources().getString(R.string.msg_network_connect_error));
                return;
            }

            if (gpsTrackerManager != null) {

                latitude = gpsTrackerManager.getLatitude();
                longitude = gpsTrackerManager.getLongitude();
                Log.e("Location", TAG + " saveServerUploadSign  GPSTrackerManager : " + latitude + "  " + longitude + "  ");
            }

            String driverMemo = edit_sign_d_f_memo.getText().toString().trim();
            if (driverMemo.equals("")) {
                Toast.makeText(this.getApplicationContext(), context.getResources().getString(R.string.msg_must_enter_memo1), Toast.LENGTH_SHORT).show();
                return;
            }

            if (!camera2.hasImage(img_sign_d_f_visit_log)) {
                Toast.makeText(this.getApplicationContext(), context.getResources().getString(R.string.msg_visit_photo_require), Toast.LENGTH_SHORT).show();
                return;
            }

            if (MemoryStatus.getAvailableInternalMemorySize() != MemoryStatus.ERROR && MemoryStatus.getAvailableInternalMemorySize() < MemoryStatus.PRESENT_BYTE) {
                AlertShow(context.getResources().getString(R.string.msg_disk_size_error));
                return;
            }

            FailedCodeResult.FailedCode code = arrayList.get(spinner_d_f_failed_reason.getSelectedItemPosition());
            String failedCode = code.getFailedCode();
            Log.e("krm0219", "DeliveryFailedUploadHelper   failedCode  " + failedCode);

            DataUtil.logEvent("button_click", TAG, com.giosis.library.util.DataUtil.requestSetUploadDeliveryData);

            new DeliveryFailedUploadHelper.Builder(DeliveryFailedActivity.this, opID, officeCode, deviceID,
                    mStrWaybillNo, failedCode, driverMemo, img_sign_d_f_visit_log,
                    MemoryStatus.getAvailableInternalMemorySize(), latitude, longitude)
                    .setOnServerEventListener(new OnServerEventListener() {

                        @Override
                        public void onPostResult() {

                            DataUtil.inProgressListPosition = 0;
                            finish();
                        }

                        @Override
                        public void onPostFailList() {
                        }
                    }).build().execute();
        } catch (Exception e) {

            Log.e("Server", TAG + "  Exception : " + e.toString());
            Toast.makeText(this.getApplicationContext(), context.getResources().getString(R.string.text_error) + " - " + e.toString(), Toast.LENGTH_SHORT).show();
        }
    }


    private void AlertShow(String msg) {

        AlertDialog.Builder alert_internet_status = new AlertDialog.Builder(this);
        alert_internet_status.setTitle(context.getResources().getString(R.string.text_warning));
        alert_internet_status.setMessage(msg);
        alert_internet_status.setPositiveButton(context.getResources().getString(R.string.button_close),
                (dialog, which) -> {
                    dialog.dismiss(); // 닫기
                    finish();
                });
        alert_internet_status.show();
    }


    // Gallery
    private void getImageFromAlbum() {
        try {

            if (!isGalleryActivate) {

                Intent intent = new Intent();
                intent.setType("image/*");
                intent.setAction(Intent.ACTION_GET_CONTENT);
                isGalleryActivate = true;
                startActivityForResult(Intent.createChooser(intent, "Select Picture"), RESULT_LOAD_IMAGE);
            }
        } catch (Exception ex) {

            isGalleryActivate = false;
            Log.i("eylee", ex.toString());
        }
    }


    // CAMERA
    private void openCamera() {

        CameraManager cameraManager = camera2.getCameraManager(this);
        cameraId = camera2.getCameraCharacteristics(cameraManager);
        Log.i("krm0219", TAG + "  openCamera " + cameraId);

        if (cameraId != null) {

            camera2.setCameraDevice(cameraManager, cameraId);
        } else {

            Toast.makeText(DeliveryFailedActivity.this, context.getResources().getString(R.string.msg_back_camera_required), Toast.LENGTH_SHORT).show();
        }
    }

    private void closeCamera() {

        camera2.closeCamera();
    }

    @Override
    public void onCameraDeviceOpened(CameraDevice cameraDevice, Size cameraSize, int rotation) {

        texture_sign_d_f_preview.setRotation(rotation);

        SurfaceTexture texture = texture_sign_d_f_preview.getSurfaceTexture();
        texture.setDefaultBufferSize(cameraSize.getWidth(), cameraSize.getHeight());
        // texture.setDefaultBufferSize(1088, 1088);
        Surface surface = new Surface(texture);

        camera2.setCaptureSessionRequest(cameraDevice, surface);
    }

    @Override
    public void onSurfaceTextureAvailable(SurfaceTexture surfaceTexture, int i, int i1) {

        openCamera();
    }

    @Override
    public void onSurfaceTextureSizeChanged(SurfaceTexture surfaceTexture, int i, int i1) {
    }

    @Override
    public boolean onSurfaceTextureDestroyed(SurfaceTexture surfaceTexture) {
        return true;
    }

    @Override
    public void onSurfaceTextureUpdated(SurfaceTexture surfaceTexture) {
    }
}

File: appMY/src/main/java/com/giosis/util/qdrive/list/delivery/DeliveryFailedActivity.kt
package com.giosis.util.qdrive.list.delivery

import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.graphics.SurfaceTexture
import android.hardware.camera2.CameraDevice
import android.hardware.camera2.CameraManager
import android.os.Bundle
import android.provider.MediaStore
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.util.Size
import android.view.Surface
import android.view.TextureView.SurfaceTextureListener
import android.view.View
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.Toast
import com.giosis.library.server.data.FailedCodeResult
import com.giosis.util.qdrive.gps.GPSTrackerManager
import com.giosis.util.qdrive.international.MyApplication
import com.giosis.util.qdrive.international.OnServerEventListener
import com.giosis.util.qdrive.international.R
import com.giosis.util.qdrive.util.*
import com.giosis.util.qdrive.util.ui.CommonActivity
import kotlinx.android.synthetic.main.activity_delivery_visit_log.*
import kotlinx.android.synthetic.main.activity_pickup_visit_log.*
import kotlinx.android.synthetic.main.top_title.*
import java.util.*


class DeliveryFailedActivity : CommonActivity(), Camera2APIs.Camera2Interface, SurfaceTextureListener {

    val tag = "DeliveryFailedActivity"
    private val context = MyApplication.getContext()

    private val userId = MyApplication.preferences.userId
    private val officeCode = MyApplication.preferences.officeCode
    private val deviceId = MyApplication.preferences.deviceUUID
    lateinit var trackingNo: String

    // Location
    private var gpsTrackerManager: GPSTrackerManager? = GPSTrackerManager(context)

    // Camera & Gallery
    private val camera2 = Camera2APIs(this)
    private var cameraId: String? = null
    private val RESULT_LOAD_IMAGE = 2000

    // Permission
    var isPermissionTrue = false
    val PERMISSION_REQUEST_CODE = 1000
    val PERMISSIONS = arrayOf(PermissionChecker.ACCESS_FINE_LOCATION, PermissionChecker.ACCESS_COARSE_LOCATION,
            PermissionChecker.READ_EXTERNAL_STORAGE, PermissionChecker.WRITE_EXTERNAL_STORAGE, PermissionChecker.CAMERA)

    //
    var arrayList: ArrayList<FailedCodeResult.FailedCode>? = null
    var failedCodeArrayList: ArrayList<String>? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        /* setShowWhenLocked(true)
         setTurnScreenOn(true)*/

        // 어플이 사용되는 동안 화면 끄지 않기
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        // Lock 상태이면 보여주지 않음. (Lock 해제해야만 보임)
        window.addFlags(WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
        // Lock 상태이면 보여줌
        window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
        // 화면 ON
        window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)

        setContentView(R.layout.activity_delivery_visit_log)


        text_top_title.text = intent.getStringExtra("title")
        trackingNo = intent.getStringExtra("trackingNo").toString()
        text_sign_d_f_tracking_no.text = trackingNo
        text_sign_d_f_receiver.text = intent.getStringExtra("receiverName")
        text_sign_d_f_sender.text = intent.getStringExtra("senderName")


        // 2020.12  Delivery Failed Code
        setFailedCode()

        layout_sign_d_f_failed_reason.setOnClickListener {

            spinner_d_f_failed_reason.performClick()
        }

        spinner_d_f_failed_reason.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(p0: AdapterView<*>?) {
            }

            override fun onItemSelected(p0: AdapterView<*>, p1: View?, p2: Int, p3: Long) {

                val reason = p0.getItemAtPosition(p2).toString()
                text_sign_d_f_failed_reason.text = reason
            }
        }

        edit_sign_d_f_memo.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(p0: Editable?) {
            }

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {

                if (99 <= edit_sign_d_f_memo.length()) {

                    Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_memo_too_long), Toast.LENGTH_SHORT).show()
                }
            }
        })



        layout_top_back.setOnClickListener {

            cancelUpload()
        }

        layout_sign_d_f_take_photo.setOnClickListener {

            if (cameraId != null) {

                camera2.takePhoto(texture_sign_d_f_preview, img_sign_d_f_visit_log)
            } else {

                Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_back_camera_required), Toast.LENGTH_SHORT).show()
            }
        }

        layout_sign_d_f_gallery.setOnClickListener {

            getImageFromGallery()
        }

        btn_sign_d_f_save.setOnClickListener {

            serverUpload()
        }

        // permission
        val checker = PermissionChecker(this)

        if (checker.lacksPermissions(*PERMISSIONS)) {

            isPermissionTrue = false
            PermissionActivity.startActivityForResult(this, PERMISSION_REQUEST_CODE, *PERMISSIONS)
            overridePendingTransition(0, 0)
        } else {

            isPermissionTrue = true
        }
    }


    private fun setFailedCode() {

        arrayList = DataUtil.getFailCode("D")

        if (arrayList == null) {

            DisplayUtil.AlertDialog(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_failed_code_error))
        } else {

            failedCodeArrayList = ArrayList()

            for (i in arrayList!!.indices) {
                val failedCode: FailedCodeResult.FailedCode = arrayList!![i]
                failedCodeArrayList!!.add(failedCode.failedString)
            }

            spinner_d_f_failed_reason.prompt = context.resources.getString(R.string.text_failed_reason)
            val failedCodeArrayAdapter = ArrayAdapter<String>(context, android.R.layout.simple_spinner_item, failedCodeArrayList!!)
            failedCodeArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            spinner_d_f_failed_reason.adapter = failedCodeArrayAdapter
        }
    }


    override fun onResume() {
        super.onResume()


        if (isPermissionTrue) {

            if (texture_sign_d_f_preview.isAvailable) {

                openCamera()
            } else {

                texture_sign_d_f_preview.surfaceTextureListener = this
            }

            // Location
            val gpsEnable = gpsTrackerManager?.enableGPSSetting()

            if (gpsEnable == true) {

                gpsTrackerManager?.GPSTrackerStart()
                Log.e(tag, " onResume  Location  :  ${gpsTrackerManager?.latitude} / ${gpsTrackerManager?.longitude}")
            } else {

                DataUtil.enableLocationSettings(this, context)
            }
        }
    }


    // Camera
    private fun openCamera() {

        val cameraManager: CameraManager = camera2.getCameraManager(this)
        cameraId = camera2.getCameraCharacteristics(cameraManager)

        if (cameraId != null) {

            camera2.setCameraDevice(cameraManager, cameraId)
        } else {

            Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_back_camera_required), Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCameraDeviceOpened(cameraDevice: CameraDevice, cameraSize: Size, rotation: Int) {

        texture_sign_d_f_preview.rotation = rotation.toFloat()

        val texture = texture_sign_d_f_preview.surfaceTexture
        texture?.setDefaultBufferSize(cameraSize.width, cameraSize.height)

        val surface = Surface(texture)
        camera2.setCaptureSessionRequest(cameraDevice, surface)
    }

    override fun onSurfaceTextureAvailable(p0: SurfaceTexture, p1: Int, p2: Int) {
        openCamera()
    }

    override fun onSurfaceTextureSizeChanged(p0: SurfaceTexture, p1: Int, p2: Int) {
    }

    override fun onSurfaceTextureUpdated(p0: SurfaceTexture) {
    }

    override fun onSurfaceTextureDestroyed(p0: SurfaceTexture): Boolean {
        return true
    }

    private fun closeCamera() {

        camera2.closeCamera()
    }

    // Gallery
    private fun getImageFromGallery() {

        val intent = Intent()
        intent.type = "image/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(Intent.createChooser(intent, "Select Picture"), RESULT_LOAD_IMAGE)
    }


    private fun cancelUpload() {

        val alertBuilder = AlertDialog.Builder(this)
        alertBuilder.setMessage(context.resources.getString(R.string.msg_delivered_sign_cancel))

        alertBuilder.setPositiveButton(context.resources.getString(R.string.button_ok)) { _, _ ->

            finish()
        }

        alertBuilder.setNegativeButton(context.resources.getString(R.string.button_cancel)) { dialogInterface, _ ->

            dialogInterface.dismiss()
        }

        alertBuilder.show()
    }


    private fun serverUpload() {

        try {

            if (!NetworkUtil.isNetworkAvailable(context)) {

                DisplayUtil.AlertDialog(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_network_connect_error))
                return
            }

            val driverMemo = edit_sign_d_f_memo.text.toString()
            if (driverMemo.isEmpty()) {

                Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_must_enter_memo1), Toast.LENGTH_SHORT).show()
                return
            }

            if (!camera2.hasImage(img_sign_d_f_visit_log)) {

                Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_visit_photo_require), Toast.LENGTH_SHORT).show()
                return
            }

            if (MemoryStatus.getAvailableInternalMemorySize() != MemoryStatus.ERROR.toLong() && MemoryStatus.getAvailableInternalMemorySize() < MemoryStatus.PRESENT_BYTE) {

                DisplayUtil.AlertDialog(this@DeliveryFailedActivity, context.resources.getString(R.string.msg_disk_size_error))
                return
            }


            var latitude = 0.0
            var longitude = 0.0
            gpsTrackerManager?.let {

                latitude = it.latitude
                longitude = it.longitude
            }

            Log.e(tag, "  Location $latitude / $longitude")


            val code: FailedCodeResult.FailedCode = arrayList!![spinner_d_f_failed_reason.selectedItemPosition]
            val failedCode: String = code.failedCode
            Log.e("krm0219", "Fail Reason Code  >  $failedCode")
            DataUtil.logEvent("button_click", tag, com.giosis.library.util.DataUtil.requestSetUploadDeliveryData)

            DeliveryFailedUploadHelper.Builder(this, userId, officeCode, deviceId,
                    trackingNo, img_sign_d_f_visit_log, failedCode, driverMemo, "RC",
                    MemoryStatus.getAvailableInternalMemorySize(), latitude, longitude)
                    .setOnServerEventListener(object : OnServerEventListener {
                        override fun onPostResult() {

                            DataUtil.inProgressListPosition = 0
                            finish()
                        }

                        override fun onPostFailList() {}
                    }).build().execute()


        } catch (e: Exception) {

            Log.e(tag, "   serverUpload  Exception : $e")
            Toast.makeText(this@DeliveryFailedActivity, context.resources.getString(R.string.text_error) + " - " + e.toString(), Toast.LENGTH_SHORT).show()
        }
    }


    override fun onBackPressed() {

        cancelUpload()
    }

    override fun onPause() {
        super.onPause()
        closeCamera()
    }

    override fun onDestroy() {
        super.onDestroy()
        DataUtil.stopGPSManager(gpsTrackerManager)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (resultCode == PermissionActivity.PERMISSIONS_GRANTED) {

                isPermissionTrue = true
            }
        } else if (requestCode == RESULT_LOAD_IMAGE && resultCode == Activity.RESULT_OK) {

            data?.let {

                try {

                    val imageUri = data.data
                    val selectedImage = MediaStore.Images.Media.getBitmap(contentResolver, imageUri)
                    val resizeImage = camera2.getResizeBitmap(selectedImage)
                    img_sign_d_f_visit_log.setImageBitmap(resizeImage)
                    img_sign_d_f_visit_log.scaleType = ImageView.ScaleType.CENTER_INSIDE
                } catch (e: Exception) {

                }
            }
        }
    }
}